### DISCUSSION TITLE
Integração com Stripe

### DISCUSSION THREAD

**Discussion by ClaudineiMonteiro** [balta-io/forum#622] [Tira Dúvidas] [answered]
Aparentemente o erro se dá por que o stripe não esta enchergando a chave publica, porém, conforme mostra na segunda imagem estou pegando a chave, e ao debugar já conferi que o valor está chegando dasecret, sabe o que pode ser feito pra funiconar?
![Stripe_Error](https://github.com/user-attachments/assets/72d65563-dbb7-4710-8703-af6afd3f158a)
![Stripe_Error_02](https://github.com/user-attachments/assets/319579f6-4069-4f1e-ae14-cc8e2e4bf38e)
![Stripe_Error](https://github.com/user-attachments/assets/5ac5fed9-1d59-44e0-abef-dc4f00759c8b)
![Stripe_Error_02](https://github.com/user-attachments/assets/2634c727-f762-40e2-bcf8-8e9da991e494)

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @ClaudineiMonteiro, como vai?

Qual erro dá? Em qual parte do código?

**Reply by ClaudineiMonteiro** [reply]
Ele consegue gravar o pedido e enviar o pedido para o stripe, porém, quando tenta redirecionar para o site de checkout do próprio stripe da o erro a seguir:
'Could not find 'checkout' ('checkout' was undefined).
Error: Could not find 'checkout' ('checkout' was undefined).
    at http://localhost:5192/_framework/blazor.webassembly.js:1:368
    at Array.forEach (<anonymous>)
    at l.findFunction (http://localhost:5192/_framework/blazor.webassembly.js:1:336)
    at w (http://localhost:5192/_framework/blazor.webassembly.js:1:5079)
    at http://localhost:5192/_framework/blazor.webassembly.js:1:2872
    at new Promise (<anonymous>)
    at b.beginInvokeJSFromDotNet (http://localhost:5192/_framework/blazor.webassembly.js:1:2835)
    at Object.vn [as invokeJSJson] (http://localhost:5192/_framework/blazor.webassembly.js:1:58849)
    at http://localhost:5192/_framework/dotnet.runtime.8.0.7.urcsr75yt5.js:3:178364
    at Tl (http://localhost:5192/_framework/dotnet.runtime.8.0.7.urcsr75yt5.js:3:179198)'
![Stripe_Error_03](https://github.com/user-attachments/assets/44fd5bdb-2c1f-41ef-83db-0f85e7f55b97)
![Stripe_Error_04](https://github.com/user-attachments/assets/d960050c-06ad-4057-a2c1-1da44c6ada3f)
![Stripe_Error_05](https://github.com/user-attachments/assets/4da68849-28df-419b-b333-f88af588614c)
O interessante é que se vou no log do stripe e pego o link que ele gera de pagamento e executo no browser separadamente ele abre o checkout do stripe e efetua o pagamento normalmente, tanto que posso consultar no saldo no stripe.

**Reply by andrebaltieri** [reply]
Parece ser um problema do seu JS na integração com o Blazor!
Pode postar o código JS aqui também por favor?

**Reply by ClaudineiMonteiro** [reply]
using Dima.Core.Handlers;
using Dima.Core.Requests.Orders;
using Dima.Core.Stripe;
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;
using MudBlazor;

namespace Dima.Web.Components;

public partial class CheckoutButtonComponent : ComponentBase
{
    #region Properties

    public bool IsBusy { get; set; } = false;

    #endregion

    #region Services

    [Inject]
    public IStripeHandler StripeHandler { get; set; } = null!;

    [Inject]
    public IOrderHandler OrderHandler { get; set; } = null!;

    [Inject]
    public IJSRuntime JsRuntime { get; set; } = null!;

    [Inject]
    public ISnackbar Snackbar { get; set; } = null!;

    #endregion

    #region Methods

    public async Task OnValidSubmitAsync()
    {
        IsBusy = true;

        var request = new CreateSessionRequest();
        try
        {
            var result = await OrderHandler.CreateOrderAsync(new CreateOrderRequest());
            if (!result.IsSuccess)
            {
                Snackbar.Add("Não foi possível registrar seu pedido", Severity.Error);
                return;
            }

            request.OrderNumber = result.Data?.Number ?? string.Empty;
        }
        catch
        {
            Snackbar.Add("Não foi possível registrar seu pedido", Severity.Error);
        }

        try
        {
            var result = await StripeHandler.CreateSessionAsync(request);
            if (result.Data is not null)
                await JsRuntime.InvokeVoidAsync("checkout", Configuration.StripePublicKey, result.Data);
            else
                Snackbar.Add("Não foi possível iniciar seu pagamento", Severity.Error);
        }
        catch (Exception ex)
        {
            Snackbar.Add($"Não foi possível iniciar seu pagamento: {ex.Message}", Severity.Error);
        }
        finally
        {
            IsBusy = false;
        }
    }

    #endregion
}

**Reply by ClaudineiMonteiro** [reply]
![image](https://github.com/user-attachments/assets/4da445a9-eb5f-4b61-9871-48fcbf68d854)

**Reply by ClaudineiMonteiro** [reply]
'Could not find 'checkout' ('checkout' was undefined).\nError: Could not find 'checkout' ('checkout' was undefined).\n    at http://localhost:5192/_framework/blazor.webassembly.js:1:368\n    at Array.forEach (<anonymous>)\n    at l.findFunction (http://localhost:5192/_framework/blazor.webassembly.js:1:336)\n    at w (http://localhost:5192/_framework/blazor.webassembly.js:1:5079)\n    at http://localhost:5192/_framework/blazor.webassembly.js:1:2872\n    at new Promise (<anonymous>)\n    at b.beginInvokeJSFromDotNet (http://localhost:5192/_framework/blazor.webassembly.js:1:2835)\n    at Object.vn [as invokeJSJson] (http://localhost:5192/_framework/blazor.webassembly.js:1:58849)\n    at http://localhost:5192/_framework/dotnet.runtime.8.0.7.urcsr75yt5.js:3:178364\n    at Tl (http://localhost:5192/_framework/dotnet.runtime.8.0.7.urcsr75yt5.js:3:179198)'

**Reply by ClaudineiMonteiro** [reply]
window.checkout = (stripePublicKey, sessionId) => {
    let stripe = Stripe(stripePublicKey);
    stripe.redirectToCheckout({
        sessionId: sessionId
    });
}

function showAlert(message) {
    alert(message);
}

**Comment by ClaudineiMonteiro** [comment]
No mesmo arquivo coloquei uma função ShowAlert, e criei uma pagina que chama esta função direto o @code.
Se Eu chamar a pagina que criai chamando está função primeiro, depois quando vou na pagina que chama o meio de pagamento ela funciona. Mas se eu chamar a pagina de pagamento primeiro ai não fuinciona

**Reply by andrebaltieri** [reply]
@ClaudineiMonteiro o erro que ele te apresenta é que não encontrou sua função `checkout`.
A função precisa estar atribuída ao Window e disponível, então o que eu normalmente faço:

1 - Crio a função no arquivo JavaScript
2 - Importo o JS na Index.html (Isto garante que a função estará disponível em todas as páginas e componentes)
3 - Invoco a função via JsRuntime do Blazor

Estes passos eu mostrei na aula e estão nos fontes do curso funcionando.
Você fez todos eles? Consegue revisar e ver se não deixou passar nada?
